// allure-mocha/dist/AllureReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllureReporter = void 0;
const path_1 = require("path");
const allure_js_commons_1 = require("allure-js-commons");
const MochaAllure_1 = require("./MochaAllure");
class AllureReporter {
    constructor(allureRuntime) {
        this.allureRuntime = allureRuntime;
        this.currentExecutable = null;
        this.cwd = process.cwd();
        this.suites = [];
        this.steps = [];
        this.runningTest = null;
    }
    getImplementation() {
        return new MochaAllure_1.MochaAllure(this, this.allureRuntime);
    }
    get currentSuite() {
        return this.suites.length > 0 ? this.suites[this.suites.length - 1] : null;
    }
    get currentStep() {
        return this.steps.length > 0 ? this.steps[this.steps.length - 1] : null;
    }
    get currentTest() {
        return this.runningTest;
    }
    set currentTest(test) {
        this.runningTest = test;
    }
    startSuite(suiteName) {
        const scope = this.currentSuite || this.allureRuntime;
        const suite = scope.startGroup(suiteName || "Global");
        this.pushSuite(suite);
    }
    endSuite() {
        if (this.currentSuite !== null) {
            if (this.currentStep !== null) {
                this.currentStep.endStep();
            }
            this.currentSuite.endGroup();
            this.popSuite();
        }
    }
    startHook(hook) {
        const suite = this.currentSuite;
        const title = hook.title;
        if (suite && title && title.includes("before")) {
            this.currentExecutable = suite.addBefore();
        }
        else if (suite && title && title.includes("after")) {
            this.currentExecutable = suite.addAfter();
        }
        if (this.currentExecutable) {
            this.currentExecutable.name = title;
        }
    }
    endHook(error) {
        if (this.currentExecutable) {
            this.currentExecutable.stage = allure_js_commons_1.Stage.FINISHED;
            if (error) {
                this.currentExecutable.status = allure_js_commons_1.Status.FAILED;
                this.currentExecutable.statusDetails = { message: error.message, trace: error.stack };
            }
            else {
                this.currentExecutable.status = allure_js_commons_1.Status.PASSED;
            }
        }
    }
    startCase(test) {
        var _a;
        if (this.currentSuite === null) {
            throw new Error("No active suite");
        }
        const testPath = (_a = test.file) === null || _a === void 0 ? void 0 : _a.replace(this.cwd, "");
        this.currentTest = this.currentSuite.startTest(test.title);
        this.currentTest.fullName = test.title;
        this.currentTest.historyId = (0, allure_js_commons_1.md5)(test.fullTitle());
        this.currentTest.stage = allure_js_commons_1.Stage.RUNNING;
        if (testPath) {
            const normalizedTestPath = (0, path_1.normalize)(testPath || "")
                .replace(/^\//, "")
                .split("/")
                .filter((item) => item !== (0, path_1.basename)(testPath));
            this.currentTest.addLabel(allure_js_commons_1.LabelName.PACKAGE, normalizedTestPath.join("."));
        }
        if (test.parent) {
            const [parentSuite, suite, ...subSuites] = this.getSuitePath(test);
            if (parentSuite) {
                this.currentTest.addLabel(allure_js_commons_1.LabelName.PARENT_SUITE, parentSuite);
            }
            if (suite) {
                this.currentTest.addLabel(allure_js_commons_1.LabelName.SUITE, suite);
            }
            if (subSuites.length > 0) {
                this.currentTest.addLabel(allure_js_commons_1.LabelName.SUB_SUITE, subSuites.join(" > "));
            }
        }
    }
    passTestCase(test) {
        if (this.currentTest === null) {
            this.startCase(test);
        }
        this.endTest(allure_js_commons_1.Status.PASSED);
    }
    pendingTestCase(test) {
        this.startCase(test);
        this.endTest(allure_js_commons_1.Status.SKIPPED, { message: "Test ignored" });
    }
    failTestCase(test, error) {
        if (this.currentTest === null) {
            this.startCase(test);
        }
        else {
            const latestStatus = this.currentTest.status;
            if (latestStatus === allure_js_commons_1.Status.FAILED || latestStatus === allure_js_commons_1.Status.BROKEN) {
                return;
            }
        }
        const status = error.name === "AssertionError" ? allure_js_commons_1.Status.FAILED : allure_js_commons_1.Status.BROKEN;
        this.endTest(status, { message: error.message, trace: error.stack });
    }
    writeAttachment(content, options) {
        return this.allureRuntime.writeAttachment(content, options);
    }
    pushStep(step) {
        this.steps.push(step);
    }
    popStep() {
        this.steps.pop();
    }
    pushSuite(suite) {
        this.suites.push(suite);
    }
    popSuite() {
        this.suites.pop();
    }
    getSuitePath(test) {
        const path = [];
        let currentSuite = test.parent;
        while (currentSuite) {
            if (currentSuite.title) {
                path.unshift(currentSuite.title);
            }
            currentSuite = currentSuite.parent;
        }
        return path;
    }
    endTest(status, details) {
        if (this.currentTest === null) {
            throw new Error("endTest while no test is running");
        }
        if (details) {
            this.currentTest.statusDetails = details;
        }
        this.currentTest.status = status;
        this.currentTest.stage = allure_js_commons_1.Stage.FINISHED;
        this.currentTest.endTest();
        this.currentTest = null;
    }
}
exports.AllureReporter = AllureReporter;
