// allure-mocha/dist/MochaAllureReporter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MochaAllureReporter = exports.allure = void 0;
const allure_js_commons_1 = require("allure-js-commons");
const Mocha = __importStar(require("mocha"));
const AllureReporter_1 = require("./AllureReporter");
const ParallelMochaAllure_1 = require("./ParallelMochaAllure");
const { EVENT_SUITE_BEGIN, EVENT_SUITE_END, EVENT_TEST_BEGIN, EVENT_TEST_END, EVENT_TEST_PASS, EVENT_TEST_FAIL, EVENT_TEST_PENDING, EVENT_HOOK_BEGIN, EVENT_HOOK_END, } = Mocha.Runner.constants;
exports.allure = new ParallelMochaAllure_1.ParallelMochaAllure();
class MochaAllureReporter extends Mocha.reporters.Base {
    constructor(runner, opts) {
        super(runner, opts);
        this.runner = runner;
        this.opts = opts;
        const { resultsDir = "allure-results" } = opts.reporterOptions || {};
        const allureConfig = Object.assign(Object.assign({}, opts.reporterOptions), { resultsDir });
        this.coreReporter = new AllureReporter_1.AllureReporter(new allure_js_commons_1.AllureRuntime(allureConfig));
        exports.allure = this.coreReporter.getImplementation();
        if (runner.linkPartialObjects) {
            runner.linkPartialObjects(true);
        }
        runner
            .on(EVENT_SUITE_BEGIN, this.onSuite.bind(this))
            .on(EVENT_SUITE_END, this.onSuiteEnd.bind(this))
            .on(EVENT_TEST_BEGIN, this.onTest.bind(this))
            .on(EVENT_TEST_PASS, this.onPassed.bind(this))
            .on(EVENT_TEST_FAIL, this.onFailed.bind(this))
            .on(EVENT_TEST_PENDING, this.onPending.bind(this))
            .on(EVENT_HOOK_BEGIN, this.onHookStart.bind(this))
            .on(EVENT_HOOK_END, this.onHookEnd.bind(this));
    }
    onSuite(suite) {
        this.coreReporter.startSuite(suite.fullTitle());
    }
    onSuiteEnd() {
        this.coreReporter.endSuite();
    }
    onTest(test) {
        this.coreReporter.startCase(test);
    }
    onPassed(test) {
        this.coreReporter.passTestCase(test);
    }
    onFailed(test, error) {
        this.coreReporter.failTestCase(test, error);
    }
    onPending(test) {
        this.coreReporter.pendingTestCase(test);
    }
    onHookStart(hook) {
        this.coreReporter.startHook(hook);
    }
    onHookEnd(hook) {
        var _a;
        this.coreReporter.endHook((_a = hook === null || hook === void 0 ? void 0 : hook.error) === null || _a === void 0 ? void 0 : _a.call(hook));
    }
}
exports.MochaAllureReporter = MochaAllureReporter;
